main.go

```diff
@@ -181,10 +182,96 @@ func (m *memo) list(n *node) {
 	}
 }
 
+type dfsStatus int
+
+const (
+	white dfsStatus = iota
+	gray
+	black
+)
+
+type dfsInfo struct {
+	Me     *class    // this class
+	Parent *class    // parent
+	D      int       // discovery time
+	F      int       // finished visiting time
+	Color  dfsStatus // WHITE (not discovered), GRAY (not visited), BLACK (done)
+}
+
+type dfsInfoList []*dfsInfo
+
+func (m *memo) DFS() dfsInfoList {
+	state := make(map[*class]*dfsInfo, len(m.classes))
+	for _, c := range m.classes {
+		state[c] = &dfsInfo{c, nil, -1, -1, white}
+	}
+	t := 0
+
+	res := make([]*dfsInfo, 0, len(m.classes))
+	for _, c := range m.classes {
+		if state[c].Color == white {
+			m.dfsVisit(c, &t, state)
+		}
+		res = append(res, state[c])
+	}
+	return res
+}
+
+func (m *memo) dfsVisit(c *class, t *int, state map[*class]*dfsInfo) {
+	*t++
+	state[c].D = *t
+	state[c].Color = gray
```

Comment:
I haven't implemented topological sort in a long time, but I seem to recall only having 2 states: visited and not visited. Perhaps I'm misremembering.